### PR TITLE
Assignment debug, documentation, and unit test

### PR DESCRIPTION
**Pull request by pchauchat**
Debugged assignment in non-trivial cases (e.g., vectorizing along axis!=0), and allowed assigning even when the former element was non-zero. Added unit tests for both cases.

Note that assignment_by_sum seems a bit redundant, especially in the case of numpy and pytorch.

Added corresponding docstrings.

I also thought of a way of documenting evolving shapes along a provided axis :
 new_x: array_like, shape=[dimension[:axis], n_samples, dimension[axis:]]
What do you think @ninamiolane @nguigs @nkoep  ?

(Also corrected a typo in plot_bch_so3)

### CODE REVIEW

File: geomstats/backend/tensorflow/__init__.py [hunk 1]
```diff
@@ -40,25 +40,33 @@ def get_mask_i_float(i, n):
     return mask_i_float
 
 
-def get_mask_float(indices, mask_shape):
+def get_mask_float(indices, mask_shape, dtype=float32):
```
Comment (ninamiolane) [inline geomstats/backend/tensorflow/__init__.py:43]: Ah, good catch! 🎉 

Actually, this could just be called `index_mask` or `indices_mask` or `mask_from_indices` and we have the `dtype` as parameter. Any chance you could update this?

File: geomstats/backend/tensorflow/__init__.py [hunk 2]
```diff
@@ -70,57 +78,211 @@ def get_mask_float(indices, mask_shape):
...
     axis: int
         Axis along which the mask is vectorized.
+    dtype: dtype
+        Type of the returned array.
 
     Returns
     -------
-    tf_mask : array
-    TODO(pchauchat): give shape of the output according to guideline
+    tf_mask : array, shape=[mask_shape[:axis], n_samples, mask_shape[axis:]]
```
Comment (ninamiolane) [inline geomstats/backend/tensorflow/__init__.py:126]: Yes for this syntax: could you update the contributing file as well?

File: tests/test_special_euclidean.py [hunk 3]
```diff
@@ -159,7 +159,6 @@ def test_random_and_belongs(self):
         base_point = self.group.random_uniform()
         result = self.group.belongs(base_point)
         expected = gs.array([True])
-        print(result, expected)
         self.assertAllClose(result, expected)
 
     # @geomstats.tests.np_and_tf_only
```
Comment (ninamiolane) [inline tests/test_special_euclidean.py:164]: Left-over comment (I know it's not your code, but could you still delete it?)

File: geomstats/backend/numpy/__init__.py [hunk 4]
```diff
@@ -140,14 +140,42 @@ def get_mask_i_float(i, n):
 
 
 def assignment(x, values, indices, axis=0):
```
Comment (ninamiolane) [inline geomstats/backend/numpy/__init__.py:142]: Could you explain why assignment is complicated in numpy? I'm just worried that we will loose a lot in performances if this part is not optimized.

For example, `x` shouldn't be copied, it should be modified, right?

File: geomstats/backend/tensorflow/__init__.py [hunk 5]
```diff
@@ -40,25 +40,33 @@ def get_mask_i_float(i, n):
...
-    tf_mask : array
+    tf_mask : array, shape=[mask_shape]
     """
     np_mask = _np.zeros(mask_shape)
-    if ndim(array(indices)) <= 1 and ndim(np_mask) == 1:
-        indices = list(indices)
+    single_index = not isinstance(indices, list)
+
+    if single_index:
+        indices = [indices]
```
Comment (ninamiolane) [inline geomstats/backend/tensorflow/__init__.py:63]: The edge case single_index is not tested in the unit tests.

File: geomstats/backend/tensorflow/__init__.py [hunk 6]
```diff
@@ -40,25 +40,33 @@ def get_mask_i_float(i, n):
...
 
-    if ndim(array(indices)) == 1 and ndim(np_mask) > 1:
+    for (nb_index, index) in enumerate(indices):
+        if not isinstance(index, tuple):
+            indices[nb_index] = (index,)
+
+    if single_index:
         if len(indices) != len(mask_shape):
```
Comment (ninamiolane) [inline geomstats/backend/tensorflow/__init__.py:70]: Ditto

File: geomstats/backend/tensorflow/__init__.py [hunk 7]
```diff
@@ -70,57 +78,211 @@ def get_mask_float(indices, mask_shape):
...
+
+    Notes
+    -----
+    If a single value is provided, it is assigned at all the indices.
+    If a list is given, it must have the same length as indices.
+    """
+    if not isinstance(values, list):
+        return assignment_single_value_by_sum(x, values, indices, axis)
+
+    else:
```
Comment (ninamiolane) [inline geomstats/backend/tensorflow/__init__.py:203]: This else is not tested.

File: geomstats/backend/numpy/__init__.py (hunk 4 above)
Comment (pchauchat) [reply geomstats/backend/numpy/__init__.py:142]: The problem is that, since Tensorflow does not allow modifying (or at least, I don't know how), if we want behaviors to be as similar as possible, this is needed. In some cases (weird, but not that uncommon), having backends which modify an object or create a new one might lead to completely different outcomes, for instance in lists.
However, there's also the possibility of specifying in the doc that behaviors differ, asking them to be cautious.

File: geomstats/backend/tensorflow/__init__.py (hunk 5 above)
Comment (pchauchat) [reply geomstats/backend/tensorflow/__init__.py:63]: True, the way it is coded, single_index is tested in `assignment_single_value`, which always call `get_mask_float` (or the vectorized version) with a list, even of length one. I'll correct it.

Comment (pchauchat) [reply geomstats/backend/tensorflow/__init__.py:63]: Or should it be kept in case get_mask_float is directly called ? But with an additional test ?

File: geomstats/backend/tensorflow/__init__.py (hunk 7 above)
Comment (pchauchat) [reply geomstats/backend/tensorflow/__init__.py:203]: I had added additonal tests on an awaiting PR. Do I bring them on this one ?

File: geomstats/backend/tensorflow/__init__.py (hunk 1 above)
Comment (pchauchat) [reply geomstats/backend/tensorflow/__init__.py:43]: Just to be sure I understand correctly:
- You want to rename the function, or one of its parameters ?
- In the first case, there would only be two parameters, `indices` and `dtype` ?

File: tests/test_special_euclidean.py (hunk 3 above)
Comment (pchauchat) [reply tests/test_special_euclidean.py:164]: It might be mine, as I was looking at which tests now worked with Pytorch and Tensorflow. I just forgot to remove them when doing the PR.
Same as above, in the next one they're gone.

File: geomstats/backend/tensorflow/__init__.py (hunk 5 above)
Comment (ninamiolane) [reply geomstats/backend/tensorflow/__init__.py:63]: If `get_mask_float` is always called with a list, then maybe we can just remove this `if`?

File: geomstats/backend/tensorflow/__init__.py (hunk 7 above)
Comment (ninamiolane) [reply geomstats/backend/tensorflow/__init__.py:203]: Let's wait for the other PR then!

File: tests/test_special_euclidean.py (hunk 3 above)
Comment (ninamiolane) [reply tests/test_special_euclidean.py:164]: OK! Can't wait for this next one :p

**Comment by pchauchat** [comment]
About the renaming and changes to reduce Deepsource errors, since they all apply to the backends on which you are, I think, currently working @nkoep, wouldn't it be better to wait until these modifications are merged and rebase on them ?

There's already a commit with renaming, but I can revert it.
